File: edge/app/services/recovery.py
from __future__ import annotations

import tempfile
from dataclasses import dataclass
from pathlib import Path

from app.backup.archiver import extract_archive
from app.backup.discovery import JobDefinition
from app.core.config import Settings, encryption_key_path
from app.core.encryption import decrypt_file, load_or_create_key
from app.services.state import StateStore
from app.services.upload import UploadClient, UploadFailure


@dataclass(slots=True)
class RecoveryError(RuntimeError):
    message: str
    status_code: int = 500

    def __str__(self) -> str:
        return self.message


class RecoveryService:
    def __init__(
        self,
        settings: Settings,
        logger,
        state_store: StateStore,
        upload_client: UploadClient,
    ) -> None:
        self.settings = settings
        self.logger = logger
        self.state_store = state_store
        self.upload_client = upload_client

    def recover_latest(self, job: JobDefinition) -> dict[str, object]:
        state = self._begin_recovery(job)
        download_path = self._temp_path(".download.tar.zst")
        decrypted_path = self._temp_path(".decrypted.tar.zst")
        try:
            download = self.upload_client.download_latest_snapshot(
                edge_id=self.settings.edge_id,
                job_name=job.job_name,
                destination=download_path,
            )
            return self._finalize_recovery(job, state, download, download_path, decrypted_path)
        except (UploadFailure, RecoveryError, Exception) as error:  # noqa: BLE001
            self._handle_recovery_error(job, state, error)
        finally:
            download_path.unlink(missing_ok=True)
            decrypted_path.unlink(missing_ok=True)

    def recover_by_fingerprint(self, job: JobDefinition, fingerprint: str) -> dict[str, object]:
        state = self._begin_recovery(job)
        download_path = self._temp_path(".download.tar.zst")
        decrypted_path = self._temp_path(".decrypted.tar.zst")
        try:
            download = self.upload_client.download_snapshot_by_fingerprint(
                edge_id=self.settings.edge_id,
                job_name=job.job_name,
                fingerprint=fingerprint,
                destination=download_path,
            )
            return self._finalize_recovery(job, state, download, download_path, decrypted_path)
        except (UploadFailure, RecoveryError, Exception) as error:  # noqa: BLE001
            self._handle_recovery_error(job, state, error)
        finally:
            download_path.unlink(missing_ok=True)
            decrypted_path.unlink(missing_ok=True)

    def _begin_recovery(self, job: JobDefinition):
        state = self.state_store.get(job.state_key)
        state.job_name = job.job_name
        state.last_status = "recovering"
        state.last_error_category = None
        state.last_error_detail = None
        self.state_store.set(job.state_key, state)
        return state

    def _finalize_recovery(self, job, state, download, download_path, decrypted_path):
        snapshot_filename = str(download["filename"])
        key = load_or_create_key(encryption_key_path())
        try:
            decrypt_file(key, download_path, decrypted_path)
        except Exception as exc:
            raise RecoveryError(
                "unable to decrypt snapshot with this Edge key", status_code=409
            ) from exc

        restored_files = extract_archive(decrypted_path, job.root_path)

        state.last_status = "recovered"
        state.last_error_category = None
        state.last_error_detail = None
        self.state_store.set(job.state_key, state)
        self.logger.info(
            "recovery_success job_name=%s path=%s snapshot=%s restored_files=%s",
            job.job_name,
            job.root_path,
            snapshot_filename,
            restored_files,
        )
        return {
            "status": "recovered",
            "job_name": job.job_name,
            "snapshot_filename": snapshot_filename,
            "restored_files": restored_files,
        }

    def _handle_recovery_error(self, job, state, error: BaseException) -> None:
        if isinstance(error, UploadFailure):
            if error.status_code == 404:
                wrapped = RecoveryError("no snapshots found on Central", status_code=404)
            elif error.category == "unauthorized":
                wrapped = RecoveryError(
                    "Central rejected the recovery request; check the shared auth token",
                    status_code=502,
                )
            elif error.category in {"network", "server", "rate_limited", "circuit_open"}:
                wrapped = RecoveryError(str(error), status_code=502)
            else:
                wrapped = RecoveryError(str(error), status_code=400)
            self._mark_failure(job, state, wrapped.message)
            self.logger.error(
                "recovery_failed job_name=%s path=%s detail=%s",
                job.job_name, job.root_path, wrapped.message,
            )
            raise wrapped from error
        if isinstance(error, RecoveryError):
            self._mark_failure(job, state, error.message)
            self.logger.error(
                "recovery_failed job_name=%s path=%s detail=%s",
                job.job_name, job.root_path, error.message,
            )
            raise error
        self._mark_failure(job, state, str(error))
        self.logger.exception("recovery_failed job_name=%s path=%s", job.job_name, job.root_path)
        raise RecoveryError(str(error), status_code=500) from error

    def _mark_failure(self, job: JobDefinition, state, detail: str) -> None:
        state.job_name = job.job_name
        state.last_status = "recovery_failed"
        state.last_error_category = "recovery"
        state.last_error_detail = detail
        self.state_store.set(job.state_key, state)

    def _temp_path(self, suffix: str) -> Path:
        self.settings.spool_dir.mkdir(parents=True, exist_ok=True)
        with tempfile.NamedTemporaryFile(
            dir=self.settings.spool_dir, delete=False, suffix=suffix
        ) as handle:
            return Path(handle.name)

File: edge/internal/services/recovery.go
package services

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"strings"

	"github.com/3to1go/edge/internal/backup"
	"github.com/3to1go/edge/internal/config"
	"github.com/3to1go/edge/internal/encryption"
)

// RecoveryError is a user-visible error from the recovery process.
type RecoveryError struct {
	Message    string
	StatusCode int
}

func (e *RecoveryError) Error() string { return e.Message }

// RecoveryResult is returned on successful recovery or preview.
type RecoveryResult struct {
	Status              string        `json:"status"`
	JobName             string        `json:"job_name"`
	RelativePath        string        `json:"relative_path,omitempty"`
	SnapshotFilename    string        `json:"snapshot_filename"`
	SnapshotFingerprint string        `json:"snapshot_fingerprint"`
	RestoredFiles       int           `json:"restored_files,omitempty"`
	Entries             interface{}   `json:"entries,omitempty"`
	TotalFiles          int           `json:"total_files,omitempty"`
	ReplaceCount        int           `json:"replace_count,omitempty"`
	AddCount            int           `json:"add_count,omitempty"`
}

// RecoveryService downloads and optionally restores archives from central.
type RecoveryService struct {
	settings   *config.Settings
	logger     *slog.Logger
	stateStore jobStateStore
	downloader snapshotDownloader
	encKey     []byte
}

func NewRecoveryService(settings *config.Settings, logger *slog.Logger, stateStore jobStateStore, downloader snapshotDownloader, encKey []byte) *RecoveryService {
	return &RecoveryService{
		settings:   settings,
		logger:     logger,
		stateStore: stateStore,
		downloader: downloader,
		encKey:     encKey,
	}
}

func (r *RecoveryService) Recover(ctx context.Context, job *backup.JobDefinition, fingerprint string) (*RecoveryResult, error) {
	state := r.beginRecovery(job)

	downloadPath := r.tempPath(".download.tar.zst")
	decryptedPath := r.tempPath(".decrypted.tar.zst")
	defer os.Remove(downloadPath)
	defer os.Remove(decryptedPath)

	filename, err := r.downloadSnapshot(ctx, job, fingerprint, downloadPath)
	if err != nil {
		return nil, r.handleError(job, &state, err)
	}

	if err := encryption.DecryptFile(r.encKey, downloadPath, decryptedPath); err != nil {
		re := &RecoveryError{Message: "unable to decrypt snapshot with this Edge key", StatusCode: 409}
		return nil, r.handleError(job, &state, re)
	}

	restored, err := backup.ExtractArchive(decryptedPath, job.RootPath)
	if err != nil {
		return nil, r.handleError(job, &state, &RecoveryError{Message: err.Error(), StatusCode: 500})
	}

	state.LastStatus = "recovered"
	state.LastErrorCategory = ""
	state.LastErrorDetail = ""
	r.stateStore.Set(job.RootPath, state)

	r.logger.Info("recovery_success",
		"job_name", job.JobName,
		"path", job.RootPath,
		"snapshot", filename,
		"restored_files", restored)

	return &RecoveryResult{
		Status:              "recovered",
		JobName:             job.JobName,
		SnapshotFilename:    filename,
		SnapshotFingerprint: fingerprintFromFilename(filename),
		RestoredFiles:       restored,
	}, nil
}

func (r *RecoveryService) Preview(ctx context.Context, job *backup.JobDefinition, fingerprint string) (*RecoveryResult, error) {
	downloadPath := r.tempPath(".download.tar.zst")
	decryptedPath := r.tempPath(".decrypted.tar.zst")
	defer os.Remove(downloadPath)
	defer os.Remove(decryptedPath)

	filename, err := r.downloadSnapshot(ctx, job, fingerprint, downloadPath)
	if err != nil {
		return nil, r.handlePreviewError(job, err)
	}

	if err := encryption.DecryptFile(r.encKey, downloadPath, decryptedPath); err != nil {
		re := &RecoveryError{Message: "unable to decrypt snapshot with this Edge key", StatusCode: 409}
		return nil, r.handlePreviewError(job, re)
	}

	preview, err := backup.ListArchiveEntries(decryptedPath, job.RootPath)
	if err != nil {
		return nil, r.handlePreviewError(job, &RecoveryError{Message: err.Error(), StatusCode: 500})
	}

	result := &RecoveryResult{
		Status:              "preview",
		JobName:             job.JobName,
		SnapshotFilename:    filename,
		SnapshotFingerprint: fingerprintFromFilename(filename),
	}
	if v, ok := preview["entries"]; ok {
		result.Entries = v
	}
	if v, ok := preview["total_files"].(int); ok {
		result.TotalFiles = v
	}
	if v, ok := preview["replace_count"].(int); ok {
		result.ReplaceCount = v
	}
	if v, ok := preview["add_count"].(int); ok {
		result.AddCount = v
	}
	return result, nil
}

func (r *RecoveryService) downloadSnapshot(ctx context.Context, job *backup.JobDefinition, fingerprint, destPath string) (string, error) {
	if fingerprint != "" {
		return r.downloader.DownloadSnapshotByFingerprint(ctx, r.settings.EdgeID, job.JobName, fingerprint, destPath)
	}
	return r.downloader.DownloadLatestSnapshot(ctx, r.settings.EdgeID, job.JobName, destPath)
}

func (r *RecoveryService) beginRecovery(job *backup.JobDefinition) JobState {
	state := r.stateStore.Get(job.RootPath)
	state.JobName = job.JobName
	state.LastStatus = "recovering"
	state.LastErrorCategory = ""
	state.LastErrorDetail = ""
	r.stateStore.Set(job.RootPath, state)
	return state
}

func (r *RecoveryService) handleError(job *backup.JobDefinition, state *JobState, err error) error {
	re := wrapRecoveryError(err, false)
	state.JobName = job.JobName
	state.LastStatus = "recovery_failed"
	state.LastErrorCategory = "recovery"
	state.LastErrorDetail = re.Message
	r.stateStore.Set(job.RootPath, *state)
	r.logger.Error("recovery_failed", "job_name", job.JobName, "path", job.RootPath, "detail", re.Message)
	return re
}

func (r *RecoveryService) handlePreviewError(job *backup.JobDefinition, err error) error {
	re := wrapRecoveryError(err, true)
	r.logger.Warn("recovery_preview_failed", "job_name", job.JobName, "path", job.RootPath, "detail", re.Message)
	return re
}

func wrapRecoveryError(err error, isPreview bool) *RecoveryError {
	if re, ok := err.(*RecoveryError); ok {
		return re
	}
	if uf, ok := err.(*UploadFailure); ok {
		switch {
		case uf.StatusCode == 404:
			return &RecoveryError{Message: "no snapshots found on Central", StatusCode: 404}
		case uf.Category == "unauthorized":
			code := 401
			if !isPreview {
				code = 502
			}
			return &RecoveryError{Message: "Central rejected the recovery request; check the Edge credential", StatusCode: code}
		case uf.Category == "network" || uf.Category == "server" || uf.Category == "rate_limited" || uf.Category == "circuit_open":
			return &RecoveryError{Message: uf.Error(), StatusCode: 502}
		default:
			return &RecoveryError{Message: uf.Error(), StatusCode: 400}
		}
	}
	return &RecoveryError{Message: err.Error(), StatusCode: 500}
}

func (r *RecoveryService) tempPath(suffix string) string {
	os.MkdirAll(r.settings.SpoolDir, 0o755)
	f, _ := os.CreateTemp(r.settings.SpoolDir, "recovery-*"+suffix)
	if f != nil {
		f.Close()
		return f.Name()
	}
	return filepath.Join(r.settings.SpoolDir, fmt.Sprintf("recovery-%d%s", os.Getpid(), suffix))
}

func fingerprintFromFilename(filename string) string {
	parts := strings.Split(filename, "__")
	if len(parts) < 3 {
		return ""
	}
	last := parts[len(parts)-1]
	last = strings.TrimSuffix(last, ".tar.zst")
	return last
}
